src/gen-ts.ts

```diff
@@ -297,9 +299,34 @@ function handleMethods(analyzerMethods: Iterable<analyzer.Method>):
     });
     m.description = method.description || '';
 
-    for (const param of method.params || []) {
-      const {type, optional, rest} = closureParamToTypeScript(param.type);
-      m.params.push(new ts.Param({name: param.name, type, optional, rest}));
+    let requiredAhead = false;
+    for (let i = (method.params || []).length - 1; i >= 0; i--) {
+      const param = method.params![i];
+      let {type, optional, rest} = closureParamToTypeScript(param.type);
+
+      if (param.defaultValue !== undefined) {
+        // Parameters with default values generally behave like optional
+        // parameters. However, unlike optional parameters, they may be
+        // followed by a required parameter, in which case the default value is
+        // set by explicitly passing undefined.
+        if (!requiredAhead) {
+          optional = true;
+        } else {
+          type = new ts.UnionType([type, ts.undefinedType]);
+        }
+      } else if (!optional) {
+        requiredAhead = true;
+      }
+
+      // Analyzer might know this is a rest parameter even if there was no
+      // JSDoc type annotation (or if it was wrong).
+      rest = rest || !!param.rest;
+      if (rest && type.kind !== 'array') {
+        // Rest types are always arrays.
```

Comment:
Maybe worth adding that the closure type is written without the Array syntax

Reply:
Done.